### ISSUE TITLE
Quickly Closing Tabs Stops Messages to Other Addons

### ISSUE THREAD

**Issue by joshuacant** [piroor/treestyletab#2532]
## Short description
When closing many tabs very quickly, messages stop being sent to addons registered to TST.

## Steps to reproduce
 1. Start Firefox with clean profile.
 2. Install TST.
 3. Install another addon that registers to TST such as "Tree Style Tab Mouse Wheel" by me.
 4. Open at least 10 tabs, 20 is better. Position the mouse cursor over the topmost tab's close button "[X]" and click as fast as you absolutely can. At least 7-8 clicks a second, slower than that and the problem doesn't seem to happen.

## Expected result
TST sends messages for all events to registered addons

## Actual result
TST stops sending messages to registered addons.

## Environment
 * Platform (OS): Windows
 * Version of Firefox: 75.0b7
 * Version (or revision) of Tree Style Tab: 3.4.7

Notes:
This issue was brought to my attention here: joshuacant/tst-wheel_and_double#24 and the original submitter thinks it's a regression between versions 3.3.6 and 3.4.1.

Thanks again, Piroor, for all you do to keep this addon alive and so incredibly well maintained.

**Comment by joanbm** [comment]
Bisected to 93b1c7b683112a5096f067a06cbac38581dea6d3

```
93b1c7b683112a5096f067a06cbac38581dea6d3 is the first bad commit
commit 93b1c7b683112a5096f067a06cbac38581dea6d3
Author: YUKI "Piro" Hiroshi <piro.outsider.reflex@gmail.com>
Date:   Tue Mar 3 19:20:38 2020 +0900

    Make more safe for delayed operation around already closed tabs

 webextensions/common/tst-api.js | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)
```

**Comment by joanbm** [comment]
Without knowing much about the code, I did some "monkey debugging" and found this:

Before 93b1c7b683112a5096f067a06cbac38581dea6d3, if I try to reproduce the problem (by closing a lot of tabs quickly), I see an apparently unhandled error (`TypeError: results is undefined`) is printed to the console. Though I can see no ill effect, this error seems to prevent `tst-api.js:spawnMessages` from running and sending a tab click message to the TSTMouseWheel add-on.

I think 93b1c7b683112a5096f067a06cbac38581dea6d3 tried to handle this unhandled error gracefully. However, with this fixed, now it doesn't crash before `tst-api.js:spawnMessages`, so it reaches this function and crashes there, when it tries to run `await browser.runtime.sendMessage(id, allowedMessage)`. Furthermore the exception handler there calls `unregisterAddon(id)` which is why the add-on stops completely working from then on.

I'm not sure what's the right way to fix this, but this seems to do the trick, by not throwing an exception when sendMessage fails due to one of those "missing tab"s:

```
diff --git a/webextensions/common/tst-api.js b/webextensions/common/tst-api.js
index eb85b1c7a..c40030ec3 100644
--- a/webextensions/common/tst-api.js
+++ b/webextensions/common/tst-api.js
@@ -953,7 +953,7 @@ function* spawnMessages(targetSet, params) {
     const allowedMessage = await sanitizeMessage(message, { id, tabProperties });
 
     try {
-      const result = await browser.runtime.sendMessage(id, allowedMessage).catch(ApiTabs.createErrorHandler());
+      const result = await browser.runtime.sendMessage(id, allowedMessage).catch(ApiTabs.createErrorHandler(ApiTabs.handleMissingTabError));
       return {
         id,
         result
```

**Comment by joanbm** [comment]
**UPDATE:** My previous patch doesn't completely fix the issue, there's a variant of the problem that still causes an error that ends up de-registering TSTMouseWheel. This can be reproduced by spamming the mouse wheel over the *last* tab so that immediately after closing, it the mouse is over either the "+" new tab button or the blank trailing area, so the next click opens a new tab which is then closed, and so on and so on.

The errors on the inspector look like this:

```
Error on sending message to tst-wheel_and_double@dontpokebadgers.com 
Object { targetType: "newtabbutton", tab: null, tabId: null, window: 1, windowId: 1, twisty: false, soundButton: false, closebox: false, button: 1, ctrlKey: false, … }
 Error: "aMessage.tab is null"
    send moz-extension://9b00e5ed-64c6-4829-8386-8255ccd4d802/common/tst-api.js:959
    spawnMessages moz-extension://9b00e5ed-64c6-4829-8386-8255ccd4d802/common/tst-api.js:981
    sendMessage moz-extension://9b00e5ed-64c6-4829-8386-8255ccd4d802/common/tst-api.js:928
    tryOperationAllowed moz-extension://9b00e5ed-64c6-4829-8386-8255ccd4d802/common/tst-api.js:880
    tryMouseOperationAllowedWithExtraContents moz-extension://9b00e5ed-64c6-4829-8386-8255ccd4d802/sidebar/mouse-event-listener.js:383
    onMouseUp moz-extension://9b00e5ed-64c6-4829-8386-8255ccd4d802/sidebar/mouse-event-listener.js:444
    wrapWithErrorHandler moz-extension://9b00e5ed-64c6-4829-8386-8255ccd4d802/sidebar/event-utils.js:191
    EventListener.handleEvent* moz-extension://9b00e5ed-64c6-4829-8386-8255ccd4d802/sidebar/mouse-event-listener.js:71
    results moz-extension://9b00e5ed-64c6-4829-8386-8255ccd4d802/extlib/EventListenerManager.js:61
    dispatchWithDetails moz-extension://9b00e5ed-64c6-4829-8386-8255ccd4d802/extlib/EventListenerManager.js:54
    dispatch moz-extension://9b00e5ed-64c6-4829-8386-8255ccd4d802/extlib/EventListenerManager.js:43
    init moz-extension://9b00e5ed-64c6-4829-8386-8255ccd4d802/sidebar/sidebar.js:239
    addAsync moz-extension://9b00e5ed-64c6-4829-8386-8255ccd4d802/common/metrics-data.js:27
    init moz-extension://9b00e5ed-64c6-4829-8386-8255ccd4d802/sidebar/sidebar.js:203
    async* moz-extension://9b00e5ed-64c6-4829-8386-8255ccd4d802/sidebar/index.js:28
```

```
Error on sending message to tst-wheel_and_double@dontpokebadgers.com 
Object { targetType: "blank", tab: null, tabId: null, window: 1, windowId: 1, twisty: false, soundButton: false, closebox: false, button: 1, ctrlKey: false, … }
 Error: "aMessage.tab is null"
    send moz-extension://9b00e5ed-64c6-4829-8386-8255ccd4d802/common/tst-api.js:959
    spawnMessages moz-extension://9b00e5ed-64c6-4829-8386-8255ccd4d802/common/tst-api.js:981
    sendMessage moz-extension://9b00e5ed-64c6-4829-8386-8255ccd4d802/common/tst-api.js:928
    tryOperationAllowed moz-extension://9b00e5ed-64c6-4829-8386-8255ccd4d802/common/tst-api.js:880
    tryMouseOperationAllowedWithExtraContents moz-extension://9b00e5ed-64c6-4829-8386-8255ccd4d802/sidebar/mouse-event-listener.js:383
    onMouseUp moz-extension://9b00e5ed-64c6-4829-8386-8255ccd4d802/sidebar/mouse-event-listener.js:444
    wrapWithErrorHandler moz-extension://9b00e5ed-64c6-4829-8386-8255ccd4d802/sidebar/event-utils.js:191
    EventListener.handleEvent* moz-extension://9b00e5ed-64c6-4829-8386-8255ccd4d802/sidebar/mouse-event-listener.js:71
    results moz-extension://9b00e5ed-64c6-4829-8386-8255ccd4d802/extlib/EventListenerManager.js:61
    dispatchWithDetails moz-extension://9b00e5ed-64c6-4829-8386-8255ccd4d802/extlib/EventListenerManager.js:54
    dispatch moz-extension://9b00e5ed-64c6-4829-8386-8255ccd4d802/extlib/EventListenerManager.js:43
    init moz-extension://9b00e5ed-64c6-4829-8386-8255ccd4d802/sidebar/sidebar.js:239
    addAsync moz-extension://9b00e5ed-64c6-4829-8386-8255ccd4d802/common/metrics-data.js:27
    init moz-extension://9b00e5ed-64c6-4829-8386-8255ccd4d802/sidebar/sidebar.js:203
    async* moz-extension://9b00e5ed-64c6-4829-8386-8255ccd4d802/sidebar/index.js:28
```

**Comment by piroor** [comment]
Thank you for investigation! Basically the auto-unregistration mechanism was introduced to avoid sending unnecessary notifications to already uninstalled addons. So the commit 078deee looks enough to solve the problem instantly.

**Comment by piroor** [comment]
And I've introduced more changes to fix fragility around `tab-mouseup` notification on such cases. I hope they changes should work with helper addons correctly.

**Comment by joanbm** [comment]
Thank you very much for taking the time to look into this issue! I'm testing those changes and will report back if I see any problem.

**Comment by protist** [comment]
I can confirm that this seems to have been fixed for me in the latest TST 3.4.8.
